src/compiler/checker.ts

```diff
@@ -9996,6 +9996,16 @@ namespace ts {
             return true;
         }
 
+        function extractIrreducible(types: Type[], flag: TypeFlags) {
+            if (every(types, t => !!(t.flags & TypeFlags.Union) && some((t as UnionType).types, tt => !!(tt.flags & flag)))) {

```

Comment:
could you add Null and Undefined to TypeFlags.IncludesMask to avoid the second check?

actually, would it be possible to add a special TypeFlags.EveryUnion that gets calculated in `addTypesToIntersection` and then added to `includes`? This uses up another TypeFlags, but is faster than trying to return two things. So..maybe not worth it.

I do notice that the two calls to extractIrreducible might iterate `typeSet` twice whenever it contains all-unions, but doesn't have any `null` or `undefined`s.

Reply:
> could you add Null and Undefined to TypeFlags.IncludesMask to avoid the second check?

Nope, because we're looking for members-of-members that are undefined-or-null. We'd need 3 brand new flags (null and undefined are already used to track weather the intersection itself contains them). We only have two unused type flags, sooo.... Yeah. This already bails early in every common scenario, the worst case where it's not needed is soemthing like a 52 element union where the first 51 contain `undefined`, but the 52nd contains `null`, which is irreducible and, notably, will trigger the "excessively large union" check in the `else` clause once the filtering fails, and cease to do _any_ more work. So the "worst case" is either the case where it needs to occur and all the work needs to be done, or where it's very close to that but not, and will likely be converted to `any`, anyway.